subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareResourceHasher.java

```diff
@@ -0,0 +1,209 @@
+            delegate.hash(snapshotContext);
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(ZipEntryContext zipEntryContext) throws IOException {
+        return lineEndingSensitivity.isCandidate(zipEntryContext) ?
+            hashContent(zipEntryContext) :
+            delegate.hash(zipEntryContext);
+    }
+
+    @Nullable
+    private HashCode hashContent(RegularFileSnapshotContext snapshotContext) {
+        NormalizedContentInfo normalizedContentInfo = collectContentInfo(new File(snapshotContext.getSnapshot().getAbsolutePath()));
+        return normalizedContentInfo.getContentType() == FileContentType.TEXT ?
+            normalizedContentInfo.getHash() :
+            delegate.hash(snapshotContext);
+    }
+
+    @Nullable
+    private HashCode hashContent(ZipEntryContext zipEntryContext) throws IOException {
+        NormalizedContentInfo normalizedContentInfo = zipEntryContext.getEntry().withInputStream(this::collectContentInfo);
+        return normalizedContentInfo.getContentType() == FileContentType.TEXT ?
+            normalizedContentInfo.getHash() :
+            delegate.hash(zipEntryContext);
+    }
+
+
+    public NormalizedContentInfo collectContentInfo(InputStream inputStream) {
+        FileContentTypeDetectingInputStream contentTypeDetectingInputStream = new FileContentTypeDetectingInputStream(inputStream);
+        ShortCircuitingInputStream<FileContentTypeDetectingInputStream> shortCircuitingInputStream = new ShortCircuitingInputStream<>(
+            contentTypeDetectingInputStream,
+            stream -> stream.getContentType() == FileContentType.BINARY
+        );
+        InputStream normalizedInputStream = new LineEndingNormalizingInputStream(shortCircuitingInputStream);
+
+        try {
+            HashCode hashCode = streamHasher.hash(normalizedInputStream);
+            return new NormalizedContentInfo(hashCode, contentTypeDetectingInputStream.getContentType(), shortCircuitingInputStream.isShortCircuited());
+        } finally {
+            closeAndIgnore(contentTypeDetectingInputStream, shortCircuitingInputStream, normalizedInputStream);
+        }
+    }
+
+    public NormalizedContentInfo collectContentInfo(File file) {
+        FileInputStream inputStream = null;
+        try {
+            inputStream = new FileInputStream(file);
+            return collectContentInfo(inputStream);
+        } catch (FileNotFoundException e) {
+            throw new UncheckedIOException(String.format("Failed to create MD5 hash for file '%s' as it does not exist.", file), e);
+        } finally {
+            closeAndIgnore(inputStream);
+        }
+    }
+
+    private static void closeAndIgnore(InputStream... inputStreams) {
+        for (InputStream inputStream : inputStreams) {
+            try {
+                if (inputStream != null) {
+                    inputStream.close();
+                }
+            } catch (IOException e) {
+                // Ignored
+            }
+        }
+    }
```

Comment:
I think these methods could be significantly simplified by throwing `IOException` further up. Then you could use `try-with-resources` everywhere.

`RegularFileHasher` should be symmetric with `ZipEntryHasher`, and throw `IOException` further up.

Reply:
I guess I'm just loathe to throw the whole hash away because the close failed.  Not sure under what circumstances that would occur, though - maybe never - so it might not be a legitimate concern.  I've aligned `RegularFileHasher` with `ZipEntryHasher`.  I also aligned the handling with what `ZipHasher` does, which is to log the error to debug and fall back to the non-normalized hash from the snapshot.  We could also just fail outright here, too, I suppose.

Reply:
After thinking this through, I decided to change this so that we just error out when we get an IOException from this method.  I could be convinced that we should fall back in this case, but I think these are likely all going to be fairly abnormal errors.